[BUG] Cannot set properties of null (setting 'peer')

### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

I am getting the error:
```
46 verbose stack TypeError: Cannot set properties of null (setting 'peer')
46 verbose stack     at visit (/usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/calc-dep-flags.js:101:54)
46 verbose stack     at visitNode (/usr/lib/node_modules/npm/node_modules/treeverse/lib/depth-descent.js:58:25)
46 verbose stack     at next (/usr/lib/node_modules/npm/node_modules/treeverse/lib/depth-descent.js:44:19)
46 verbose stack     at depth (/usr/lib/node_modules/npm/node_modules/treeverse/lib/depth-descent.js:83:10)
46 verbose stack     at depth (/usr/lib/node_modules/npm/node_modules/treeverse/lib/depth.js:27:12)
46 verbose stack     at unsetFlag (/usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/calc-dep-flags.js:96:5)
46 verbose stack     at /usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/calc-dep-flags.js:63:7
46 verbose stack     at Map.forEach (<anonymous>)
46 verbose stack     at calcDepFlagsStep (/usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/calc-dep-flags.js:41:17)
46 verbose stack     at visit (/usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/calc-dep-flags.js:12:20)
```
I've then checked `/usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/calc-dep-flags.js:101:54)` and added a `console.error({node});` to line 101, to get more insight. It seems like `target` is `null`, I've put the last three nodes below.

I've created a mwe [here](https://github.com/dderjoel/mwe-measuresuite). To reproduce, execute `npm i` and then try to install `npm i measuresuite@latest` again. It works well for the first install and any subsequent install, iff `node_modules` are deleted beforehand.
I am the maintainer of `measuresuite`, which is a C-library with a gyp file. The native code it build into `./build/Release/measuresuite` which I am trying to reference from `measuresuite-native-module`. c.f. [package.json:47](https://github.com/0xADE1A1DE/MeasureSuite/blob/8034055b90b57e01f20683c2c9d352bd48c62886/package.json#L47)

Possibly it is not an issue with npmcli (although it seems related like [here](https://github.com/npm/cli/issues/3711)), but with how I am packaging `measuresuite`, but I could not find any better documentation of how to package a native module. (and it used to work fine just like that).
I'm more than happy to provide more information or data if needed, but I am lost and don't know where to debug / which data is needed. 


Last three nodes:
```js
{
  node: ArboristLink {
    name: 'measuresuite-native-module',
    location: 'node_modules/measuresuite-native-module',
    path: '/home/joel/dev/6mwe/mwe-measuresuite/node_modules/measuresuite-native-module',
    realpath: '/home/joel/dev/6mwe/mwe-measuresuite/node_modules/measuresuite/build/Release/measuresuite.node',
    resolved: 'file:measuresuite/build/Release/measuresuite.node',
    optional: true,
    errors: [ [Object] ],
    edgesIn: Set(1) {
      { node_modules/measuresuite prod measuresuite-native-module@file:./build/Release/measuresuite.node }
    },
    target: ArboristNode {
      name: 'measuresuite.node',
      location: 'node_modules/measuresuite/build/Release/measuresuite.node',
      path: '/home/joel/dev/6mwe/mwe-measuresuite/node_modules/measuresuite/build/Release/measuresuite.node',
      dev: true,
      optional: true,
      errors: [Array]
    }
  }
}
{
  node: ArboristLink {
    name: 'measuresuite-native-module',
    location: 'node_modules/measuresuite-native-module',
    path: '/home/joel/dev/6mwe/mwe-measuresuite/node_modules/measuresuite-native-module',
    realpath: '/home/joel/dev/6mwe/mwe-measuresuite/node_modules/measuresuite/build/Release/measuresuite.node',
    resolved: 'file:measuresuite/build/Release/measuresuite.node',
    errors: [ [Object] ],
    edgesIn: Set(1) {
      { node_modules/measuresuite prod measuresuite-native-module@file:./build/Release/measuresuite.node }
    },
    target: ArboristNode {
      name: 'measuresuite.node',
      location: 'node_modules/measuresuite/build/Release/measuresuite.node',
      path: '/home/joel/dev/6mwe/mwe-measuresuite/node_modules/measuresuite/build/Release/measuresuite.node',
      optional: true,
      errors: [Array]
    }
  }
}
{⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂⠂) ⠼ idealTree: timing arborist:ctor Completed in 1ms
  node: ArboristLink {
    name: 'measuresuite-native-module',
    packageName: 'measuresuite.node',
    location: 'node_modules/measuresuite-native-module',
    path: '/home/joel/dev/6mwe/mwe-measuresuite/node_modules/measuresuite-native-module',
    realpath: '/home/joel/dev/6mwe/mwe-measuresuite/node_modules/measuresuite/build/Release/measuresuite.node',
    resolved: 'file:measuresuite/build/Release/measuresuite.node',
    dev: true,
    optional: true,
    edgesIn: Set(1) {
      { node_modules/measuresuite prod measuresuite-native-module@file:./build/Release/measuresuite.node }
    },
    target: null
  }
}
```



### Expected Behavior

That `npm i measuresuite@latest`  does not fail.

### Steps To Reproduce

```bash
git clone https://github.com/dderjoel/mwe-measuresuite.git
cd mwe-measuresuite
npm i
npm i measuresuite@latest
```

### Environment

- npm: 9.8.0
- Node.js: v20.3.1
- OS Name: ArchLinux: (`uname -a`: Linux arch4 6.3.9-arch1-1 #1 SMP PREEMPT_DYNAMIC Wed, 21 Jun 2023 20:46:20 +0000 x86_64 GNU/Linux)
- System Model Name: Lenovo T490
- npm config: `npm config ls`:
 ```
; "user" config from /home/joel/.npmrc

//registry.npmjs.org/:_authToken = (protected) 

; node bin location = /usr/bin/node
; node version = v20.3.1
; npm local prefix = /home/joel/dev/6mwe/mwe-measuresuite
; npm version = 9.7.2
; cwd = /home/joel/dev/6mwe/mwe-measuresuite
; HOME = /home/joel
```
 `npm config ls -l`:

```
; "default" config from default values

_auth = (protected) 
access = null 
all = false 
allow-same-version = false 
also = null 
audit = true 
audit-level = null 
auth-type = "web" 
before = null 
bin-links = true 
browser = null 
ca = null 
cache = "/home/joel/.npm" 
cache-max = null 
cache-min = 0 
cafile = null 
call = "" 
cert = null 
ci-name = null 
cidr = null 
color = true 
commit-hooks = true 
depth = null 
description = true 
dev = false 
diff = [] 
diff-dst-prefix = "b/" 
diff-ignore-all-space = false 
diff-name-only = false 
diff-no-prefix = false 
diff-src-prefix = "a/" 
diff-text = false 
diff-unified = 3 
dry-run = false 
editor = "nvim" 
engine-strict = false 
fetch-retries = 2 
fetch-retry-factor = 10 
fetch-retry-maxtimeout = 60000 
fetch-retry-mintimeout = 10000 
fetch-timeout = 300000 
force = false 
foreground-scripts = false 
format-package-lock = true 
fund = true 
git = "git" 
git-tag-version = true 
global = false 
global-style = false 
globalconfig = "/usr/etc/npmrc" 
heading = "npm" 
https-proxy = null 
if-present = false 
ignore-scripts = false 
include = [] 
include-staged = false 
include-workspace-root = false 
init-author-email = "" 
init-author-name = "" 
init-author-url = "" 
init-license = "ISC" 
init-module = "/home/joel/.npm-init.js" 
init-version = "1.0.0" 
init.author.email = "" 
init.author.name = "" 
init.author.url = "" 
init.license = "ISC" 
init.module = "/home/joel/.npm-init.js" 
init.version = "1.0.0" 
install-links = false 
install-strategy = "hoisted" 
json = false 
key = null 
legacy-bundling = false 
legacy-peer-deps = false 
link = false 
local-address = null 
location = "user" 
lockfile-version = null 
loglevel = "notice" 
logs-dir = null 
logs-max = 10 
; long = false ; overridden by cli
maxsockets = 15 
message = "%s" 
metrics-registry = "https://registry.npmjs.org/" 
node-options = null 
noproxy = [""] 
npm-version = "9.7.2" 
offline = false 
omit = [] 
omit-lockfile-registry-resolved = false 
only = null 
optional = null 
otp = null 
pack-destination = "." 
package = [] 
package-lock = true 
package-lock-only = false 
parseable = false 
prefer-dedupe = false 
prefer-offline = false 
prefer-online = false 
prefix = "/usr" 
preid = "" 
production = null 
progress = true 
provenance = false 
provenance-file = null 
proxy = null 
read-only = false 
rebuild-bundle = true 
registry = "https://registry.npmjs.org/" 
replace-registry-host = "npmjs" 
save = true 
save-bundle = false 
save-dev = false 
save-exact = false 
save-optional = false 
save-peer = false 
save-prefix = "^" 
save-prod = false 
scope = "" 
script-shell = null 
searchexclude = "" 
searchlimit = 20 
searchopts = "" 
searchstaleness = 900 
shell = "/bin/zsh" 
shrinkwrap = true 
sign-git-commit = false 
sign-git-tag = false 
strict-peer-deps = false 
strict-ssl = true 
tag = "latest" 
tag-version-prefix = "v" 
timing = false 
tmp = "/tmp" 
umask = 0 
unicode = true 
update-notifier = true 
usage = false 
user-agent = "npm/{npm-version} node/{node-version} {platform} {arch} workspaces/{workspaces} {ci}" 
userconfig = "/home/joel/.npmrc" 
version = false 
versions = false 
viewer = "man" 
which = null 
workspace = [] 
workspaces = null 
workspaces-update = true 
yes = null 

; "user" config from /home/joel/.npmrc

//registry.npmjs.org/:_authToken = (protected) 

; "cli" config from command line options

long = true

```

I just encountered this problem and would like to add some information for my personal workaround.

My project has some private scoped packages as dependencies that I have published on GitHub.
When I now use either `npm link` or local paths in the package.json and I do an `npm update` and/or `npm install` I encounter the `npm ERR! Cannot set properties of null (setting 'peer')` error.

After some time I discovered that I can prevent the error when I remove the `.npmrc` config for the org scope which was part of the recommended configuration from the GitHub documentation on private packages.

`@<myorg>:registry=https://npm.pkg.github.com/`

When I remove that line I can use `npm link` or file paths fine; although I had to link all scoped dependencies at the same time.

--- a/workspaces/arborist/lib/calc-dep-flags.js
+++ b/workspaces/arborist/lib/calc-dep-flags.js
@@ -101,13 +101,13 @@ const unsetFlag = (node, flag) => {
       tree: node,
       visit: node => {
         node.extraneous = node[flag] = false
-        if (node.isLink) {
+        if (node.isLink && node.target) {
           node.target.extraneous = node.target[flag] = false
         }
       },
       getChildren: node => {
         const children = []
-        const targetNode = node.isLink ? node.target : node
+        const targetNode = node.isLink && node.target ? node.target : node;
         for (const edge of targetNode.edgesOut.values()) {
           if (
             edge.to &&

--- a/workspaces/arborist/lib/calc-dep-flags.js
+++ b/workspaces/arborist/lib/calc-dep-flags.js
@@ -31,10 +31,6 @@ const calcDepFlagsStep = (node) => {
 
   // for links, map their hierarchy appropriately
   if (node.isLink) {
-    // node.target can be null, we check to ensure it's not null before proceeding
-    if (node.target == null) {
-      return node
-    }
     node.target.dev = node.dev
     node.target.optional = node.optional
     node.target.devOptional = node.devOptional
@@ -102,28 +98,14 @@ const unsetFlag = (node, flag) => {
       visit: node => {
         node.extraneous = node[flag] = false
         if (node.isLink) {
-          // node.target can be null, we check to ensure it's not null before proceeding
-          if (node.target == null) {
-            return []
-          } else {
-            node.target.extraneous = node.target[flag] = false
-          }
+          node.target.extraneous = node.target[flag] = false
         }
       },
       getChildren: node => {
-        if (node.isLink && node.target == null) {
-          return []
-        }
         const children = []
-        const targetNode = node.isLink ? node.target : node
-        if (targetNode == null) {
-          return []
-        }
-        for (const edge of targetNode.edgesOut.values()) {
-          if (
-            edge.to &&
-            edge.to[flag] &&
-            ((flag !== 'peer' && edge.type === 'peer') || edge.type === 'prod')
+        for (const edge of node.target.edgesOut.values()) {
+          if (edge.to && edge.to[flag] &&
+            (flag !== 'peer' && edge.type === 'peer' || edge.type === 'prod')
           ) {
             children.push(edge.to)
           }

--- a/workspaces/arborist/lib/calc-dep-flags.js
+++ b/workspaces/arborist/lib/calc-dep-flags.js
@@ -107,7 +107,7 @@ const unsetFlag = (node, flag) => {
       },
       getChildren: node => {
         const children = []
-        const targetNode = node.isLink && node.target ? node.target : node;
+        const targetNode = node.isLink && node.target ? node.target : node
         for (const edge of targetNode.edgesOut.values()) {
           if (
             edge.to &&

--- a/workspaces/arborist/lib/calc-dep-flags.js
+++ b/workspaces/arborist/lib/calc-dep-flags.js
@@ -31,6 +31,10 @@ const calcDepFlagsStep = (node) => {
 
   // for links, map their hierarchy appropriately
   if (node.isLink) {
+    // node.target can be null, we check to ensure it's not null before proceeding
+    if (node.target == null) {
+      return node
+    }
     node.target.dev = node.dev
     node.target.optional = node.optional
     node.target.devOptional = node.devOptional
@@ -103,9 +107,12 @@ const unsetFlag = (node, flag) => {
       },
       getChildren: node => {
         const children = []
-        for (const edge of node.target.edgesOut.values()) {
-          if (edge.to && edge.to[flag] &&
-            (flag !== 'peer' && edge.type === 'peer' || edge.type === 'prod')
+        const targetNode = node.isLink ? node.target : node
+        for (const edge of targetNode.edgesOut.values()) {
+          if (
+            edge.to &&
+            edge.to[flag] &&
+            ((flag !== 'peer' && edge.type === 'peer') || edge.type === 'prod')
           ) {
             children.push(edge.to)
           }

--- a/workspaces/arborist/lib/calc-dep-flags.js
+++ b/workspaces/arborist/lib/calc-dep-flags.js
@@ -1,119 +1,138 @@
-const { depth } = require('treeverse')
+const { depth } = require("treeverse");
 
 const calcDepFlags = (tree, resetRoot = true) => {
   if (resetRoot) {
-    tree.dev = false
-    tree.optional = false
-    tree.devOptional = false
-    tree.peer = false
+    tree.dev = false;
+    tree.optional = false;
+    tree.devOptional = false;
+    tree.peer = false;
   }
   const ret = depth({
     tree,
-    visit: node => calcDepFlagsStep(node),
-    filter: node => node,
+    visit: (node) => calcDepFlagsStep(node),
+    filter: (node) => node,
     getChildren: (node, tree) =>
-      [...tree.edgesOut.values()].map(edge => edge.to),
-  })
-  return ret
-}
+      [...tree.edgesOut.values()].map((edge) => edge.to),
+  });
+  return ret;
+};
 
 const calcDepFlagsStep = (node) => {
   // This rewalk is necessary to handle cases where devDep and optional
   // or normal dependency graphs overlap deep in the dep graph.
   // Since we're only walking through deps that are not already flagged
   // as non-dev/non-optional, it's typically a very shallow traversal
-  node.extraneous = false
-  resetParents(node, 'extraneous')
-  resetParents(node, 'dev')
-  resetParents(node, 'peer')
-  resetParents(node, 'devOptional')
-  resetParents(node, 'optional')
+  node.extraneous = false;
+  resetParents(node, "extraneous");
+  resetParents(node, "dev");
+  resetParents(node, "peer");
+  resetParents(node, "devOptional");
+  resetParents(node, "optional");
 
   // for links, map their hierarchy appropriately
   if (node.isLink) {
-    node.target.dev = node.dev
-    node.target.optional = node.optional
-    node.target.devOptional = node.devOptional
-    node.target.peer = node.peer
-    return calcDepFlagsStep(node.target)
+    // node.target can be null, we check to ensure it's not null before proceeding
+    if (node.target == null) {
+      return node;
+    }
+    node.target.dev = node.dev;
+    node.target.optional = node.optional;
+    node.target.devOptional = node.devOptional;
+    node.target.peer = node.peer;
+    return calcDepFlagsStep(node.target);
   }
 
   node.edgesOut.forEach(({ peer, optional, dev, to }) => {
     // if the dep is missing, then its flags are already maximally unset
     if (!to) {
-      return
+      return;
     }
 
     // everything with any kind of edge into it is not extraneous
-    to.extraneous = false
+    to.extraneous = false;
 
     // devOptional is the *overlap* of the dev and optional tree.
     // however, for convenience and to save an extra rewalk, we leave
     // it set when we are in *either* tree, and then omit it from the
     // package-lock if either dev or optional are set.
-    const unsetDevOpt = !node.devOptional && !node.dev && !node.optional && !dev && !optional
+    const unsetDevOpt =
+      !node.devOptional && !node.dev && !node.optional && !dev && !optional;
 
     // if we are not in the devOpt tree, then we're also not in
     // either the dev or opt trees
-    const unsetDev = unsetDevOpt || !node.dev && !dev
-    const unsetOpt = unsetDevOpt || !node.optional && !optional
-    const unsetPeer = !node.peer && !peer
+    const unsetDev = unsetDevOpt || (!node.dev && !dev);
+    const unsetOpt = unsetDevOpt || (!node.optional && !optional);
+    const unsetPeer = !node.peer && !peer;
 
     if (unsetPeer) {
-      unsetFlag(to, 'peer')
+      unsetFlag(to, "peer");
     }
 
     if (unsetDevOpt) {
-      unsetFlag(to, 'devOptional')
+      unsetFlag(to, "devOptional");
     }
 
     if (unsetDev) {
-      unsetFlag(to, 'dev')
+      unsetFlag(to, "dev");
     }
 
     if (unsetOpt) {
-      unsetFlag(to, 'optional')
+      unsetFlag(to, "optional");
     }
-  })
+  });
 
-  return node
-}
+  return node;
+};
 
 const resetParents = (node, flag) => {
   if (node[flag]) {
-    return
+    return;
   }
 
   for (let p = node; p && (p === node || p[flag]); p = p.resolveParent) {
-    p[flag] = false
+    p[flag] = false;
   }
-}
+};
 
 // typically a short walk, since it only traverses deps that have the flag set.
 const unsetFlag = (node, flag) => {
   if (node[flag]) {
-    node[flag] = false
+    node[flag] = false;
     depth({
       tree: node,
-      visit: node => {
-        node.extraneous = node[flag] = false
+      visit: (node) => {
+        node.extraneous = node[flag] = false;
         if (node.isLink) {
-          node.target.extraneous = node.target[flag] = false
-        }
-      },
-      getChildren: node => {
-        const children = []
-        for (const edge of node.target.edgesOut.values()) {
-          if (edge.to && edge.to[flag] &&
-            (flag !== 'peer' && edge.type === 'peer' || edge.type === 'prod')
-          ) {
-            children.push(edge.to)
+          // node.target can be null, we check to ensure it's not null before proceeding
+          if (node.target == null) {
+            return [];
+          } else {
+            node.target.extraneous = node.target[flag] = false;
           }
         }
-        return children
       },
-    })
+      getChildren: (node) => {
+        if (node.isLink && node.target == null) {
+          return [];
+        }
+        const children = [];
+        const targetNode = node.isLink ? node.target : node;
+        if (targetNode == null) {
+          return [];
+        }
+        for (const edge of targetNode.edgesOut.values()) {
+          if (
+            edge.to &&
+            edge.to[flag] &&
+            ((flag !== "peer" && edge.type === "peer") || edge.type === "prod")
+          ) {
+            children.push(edge.to);
+          }
+        }
+        return children;
+      },
+    });
   }
-}
+};
 
-module.exports = calcDepFlags
+module.exports = calcDepFlags;

--- a/workspaces/arborist/lib/calc-dep-flags.js
+++ b/workspaces/arborist/lib/calc-dep-flags.js
@@ -1,138 +1,137 @@
-const { depth } = require("treeverse");
+const { depth } = require('treeverse')
 
 const calcDepFlags = (tree, resetRoot = true) => {
   if (resetRoot) {
-    tree.dev = false;
-    tree.optional = false;
-    tree.devOptional = false;
-    tree.peer = false;
+    tree.dev = false
+    tree.optional = false
+    tree.devOptional = false
+    tree.peer = false
   }
   const ret = depth({
     tree,
-    visit: (node) => calcDepFlagsStep(node),
-    filter: (node) => node,
+    visit: node => calcDepFlagsStep(node),
+    filter: node => node,
     getChildren: (node, tree) =>
-      [...tree.edgesOut.values()].map((edge) => edge.to),
-  });
-  return ret;
-};
+      [...tree.edgesOut.values()].map(edge => edge.to),
+  })
+  return ret
+}
 
 const calcDepFlagsStep = (node) => {
   // This rewalk is necessary to handle cases where devDep and optional
   // or normal dependency graphs overlap deep in the dep graph.
   // Since we're only walking through deps that are not already flagged
   // as non-dev/non-optional, it's typically a very shallow traversal
-  node.extraneous = false;
-  resetParents(node, "extraneous");
-  resetParents(node, "dev");
-  resetParents(node, "peer");
-  resetParents(node, "devOptional");
-  resetParents(node, "optional");
+  node.extraneous = false
+  resetParents(node, 'extraneous')
+  resetParents(node, 'dev')
+  resetParents(node, 'peer')
+  resetParents(node, 'devOptional')
+  resetParents(node, 'optional')
 
   // for links, map their hierarchy appropriately
   if (node.isLink) {
     // node.target can be null, we check to ensure it's not null before proceeding
     if (node.target == null) {
-      return node;
+      return node
     }
-    node.target.dev = node.dev;
-    node.target.optional = node.optional;
-    node.target.devOptional = node.devOptional;
-    node.target.peer = node.peer;
-    return calcDepFlagsStep(node.target);
+    node.target.dev = node.dev
+    node.target.optional = node.optional
+    node.target.devOptional = node.devOptional
+    node.target.peer = node.peer
+    return calcDepFlagsStep(node.target)
   }
 
   node.edgesOut.forEach(({ peer, optional, dev, to }) => {
     // if the dep is missing, then its flags are already maximally unset
     if (!to) {
-      return;
+      return
     }
 
     // everything with any kind of edge into it is not extraneous
-    to.extraneous = false;
+    to.extraneous = false
 
     // devOptional is the *overlap* of the dev and optional tree.
     // however, for convenience and to save an extra rewalk, we leave
     // it set when we are in *either* tree, and then omit it from the
     // package-lock if either dev or optional are set.
-    const unsetDevOpt =
-      !node.devOptional && !node.dev && !node.optional && !dev && !optional;
+    const unsetDevOpt = !node.devOptional && !node.dev && !node.optional && !dev && !optional
 
     // if we are not in the devOpt tree, then we're also not in
     // either the dev or opt trees
-    const unsetDev = unsetDevOpt || (!node.dev && !dev);
-    const unsetOpt = unsetDevOpt || (!node.optional && !optional);
-    const unsetPeer = !node.peer && !peer;
+    const unsetDev = unsetDevOpt || !node.dev && !dev
+    const unsetOpt = unsetDevOpt || !node.optional && !optional
+    const unsetPeer = !node.peer && !peer
 
     if (unsetPeer) {
-      unsetFlag(to, "peer");
+      unsetFlag(to, 'peer')
     }
 
     if (unsetDevOpt) {
-      unsetFlag(to, "devOptional");
+      unsetFlag(to, 'devOptional')
     }
 
     if (unsetDev) {
-      unsetFlag(to, "dev");
+      unsetFlag(to, 'dev')
     }
 
     if (unsetOpt) {
-      unsetFlag(to, "optional");
+      unsetFlag(to, 'optional')
     }
-  });
+  })
 
-  return node;
-};
+  return node
+}
 
 const resetParents = (node, flag) => {
   if (node[flag]) {
-    return;
+    return
   }
 
   for (let p = node; p && (p === node || p[flag]); p = p.resolveParent) {
-    p[flag] = false;
+    p[flag] = false
   }
-};
+}
 
 // typically a short walk, since it only traverses deps that have the flag set.
 const unsetFlag = (node, flag) => {
   if (node[flag]) {
-    node[flag] = false;
+    node[flag] = false
     depth({
       tree: node,
-      visit: (node) => {
-        node.extraneous = node[flag] = false;
+      visit: node => {
+        node.extraneous = node[flag] = false
         if (node.isLink) {
           // node.target can be null, we check to ensure it's not null before proceeding
           if (node.target == null) {
-            return [];
+            return []
           } else {
-            node.target.extraneous = node.target[flag] = false;
+            node.target.extraneous = node.target[flag] = false
           }
         }
       },
-      getChildren: (node) => {
+      getChildren: node => {
         if (node.isLink && node.target == null) {
-          return [];
+          return []
         }
-        const children = [];
-        const targetNode = node.isLink ? node.target : node;
+        const children = []
+        const targetNode = node.isLink ? node.target : node
         if (targetNode == null) {
-          return [];
+          return []
         }
         for (const edge of targetNode.edgesOut.values()) {
           if (
             edge.to &&
             edge.to[flag] &&
-            ((flag !== "peer" && edge.type === "peer") || edge.type === "prod")
+            ((flag !== 'peer' && edge.type === 'peer') || edge.type === 'prod')
           ) {
-            children.push(edge.to);
+            children.push(edge.to)
           }
         }
-        return children;
+        return children
       },
-    });
+    })
   }
-};
+}
 
-module.exports = calcDepFlags;
+module.exports = calcDepFlags

--- a/workspaces/arborist/test/calc-dep-flags.js
+++ b/workspaces/arborist/test/calc-dep-flags.js
@@ -266,6 +266,7 @@ t.test('set parents to not extraneous when visiting', t => {
 })
 
 t.test('check null target in link', async t => {
+
   const root = new Link({
     path: '/some/path',
     realpath: '/some/path',
@@ -273,33 +274,6 @@ t.test('check null target in link', async t => {
       dependencies: { foo: '' },
     },
   });
-
-  if (root.target == null) {
-    t.equal(root.target, null, 'root.target should be null');
-    t.equal(root, root, 'should return the node itself if target is null');
-  } else {
-    root.target.peer = true;
-  }
-
-  const nonNullTarget = new Node({
-    name: "notNull",
-    path: '/some/path/node',
-  })
-
-  const rootWithNonNullTarget = new Link({
-    name: "notNull",
-    path: '/some/path/non/null',
-    pkg: {
-      dependencies: { foo: '' },
-    },
-    target: nonNullTarget
-  });
-  if (rootWithNonNullTarget.target == null) {
-    t.equal(rootWithNonNullTarget.target, null, 'rootWithNonNullTarget.target should be null');
-    t.equal(rootWithNonNullTarget, rootWithNonNullTarget, 'should return the node itself if target is null');
-  } else {
-    rootWithNonNullTarget.target.peer = true;
-    t.equal(rootWithNonNullTarget.target.peer, true, 'rootWithNonNullTarget.target.peer should be true');
-  }
+  t.doesNotThrow(()=> calcDepFlags(root))
   t.end();
 });

--- a/workspaces/arborist/test/calc-dep-flags.js
+++ b/workspaces/arborist/test/calc-dep-flags.js
@@ -265,8 +265,7 @@ t.test('set parents to not extraneous when visiting', t => {
   t.end()
 })
 
-t.test('check null target in link', async t => {
-
+t.test('check null target in link', async t => {  
   const root = new Link({
     path: '/some/path',
     realpath: '/some/path',
@@ -274,6 +273,7 @@ t.test('check null target in link', async t => {
       dependencies: { foo: '' },
     },
   });
-  t.doesNotThrow(()=> calcDepFlags(root))
+  t.doesNotThrow(() => calcDepFlags(root))
+  t.doesNotThrow(() => calcDepFlags(root, false))
   t.end();
 });

--- a/workspaces/arborist/test/calc-dep-flags.js
+++ b/workspaces/arborist/test/calc-dep-flags.js
@@ -265,15 +265,15 @@ t.test('set parents to not extraneous when visiting', t => {
   t.end()
 })
 
-t.test('check null target in link', async t => {  
+t.test('check null target in link', async t => {
   const root = new Link({
     path: '/some/path',
     realpath: '/some/path',
     pkg: {
       dependencies: { foo: '' },
     },
-  });
+  })
   t.doesNotThrow(() => calcDepFlags(root))
   t.doesNotThrow(() => calcDepFlags(root, false))
-  t.end();
-});
+  t.end()
+})

--- a/workspaces/arborist/test/calc-dep-flags.js
+++ b/workspaces/arborist/test/calc-dep-flags.js
@@ -264,3 +264,42 @@ t.test('set parents to not extraneous when visiting', t => {
   t.equal(bazLink.devOptional, false, 'bazlink not devOptional')
   t.end()
 })
+
+t.test('check null target in link', async t => {
+  const root = new Link({
+    path: '/some/path',
+    realpath: '/some/path',
+    pkg: {
+      dependencies: { foo: '' },
+    },
+  });
+
+  if (root.target == null) {
+    t.equal(root.target, null, 'root.target should be null');
+    t.equal(root, root, 'should return the node itself if target is null');
+  } else {
+    root.target.peer = true;
+  }
+
+  const nonNullTarget = new Node({
+    name: "notNull",
+    path: '/some/path/node',
+  })
+
+  const rootWithNonNullTarget = new Link({
+    name: "notNull",
+    path: '/some/path/non/null',
+    pkg: {
+      dependencies: { foo: '' },
+    },
+    target: nonNullTarget
+  });
+  if (rootWithNonNullTarget.target == null) {
+    t.equal(rootWithNonNullTarget.target, null, 'rootWithNonNullTarget.target should be null');
+    t.equal(rootWithNonNullTarget, rootWithNonNullTarget, 'should return the node itself if target is null');
+  } else {
+    rootWithNonNullTarget.target.peer = true;
+    t.equal(rootWithNonNullTarget.target.peer, true, 'rootWithNonNullTarget.target.peer should be true');
+  }
+  t.end();
+});

